src/clustering/image_clustering.h

```diff
@@ -38,36 +38,85 @@
 #include <queue>
 #include <unordered_map>
 #include <vector>
+#include <mutex>
 
 #include "clustering/cluster.h"
 #include "graph/graph.h"
 #include "util/hash.h"
 #include "util/timer.h"
 #include "util/types.h"
+#include "util/threading.h"
 
 using namespace colmap;
 
 namespace DAGSfM {
 
 using Edges = std::unordered_map<ImagePair, int>;
-struct ImageCluster {
+class ImageCluster {
+ private:
+  ImageCluster(const ImageCluster& rhs, const std::lock_guard<std::mutex> &) 
+  : cluster_id(rhs.cluster_id), image_ids_(rhs.image_ids_), edges_(rhs.edges_),
```

Comment:
It would be better to put each member variables in separate lines:
```C++
cluster_id(rhs.cluster_id),
image_ids_(rhs.image_ids_),
edges_(rhs.edges_),
...
```